src/array/cpu/spmat_op_impl_coo.cc

```diff
@@ -0,0 +1,497 @@
+
+  std::vector<IdType> indices;
+  std::vector<DType> data;
+
+  for (int64_t i = 0; i < coo_row->shape[0]; ++i) {
+    if (coo_row_data[i] == row) {
+      indices.push_back(coo_col_data[i]);
+      data.push_back(coo_data[i]);
+    }
+  }
+
+  return std::make_pair(NDArray::FromVector(data), NDArray::FromVector(indices));
+}
+
+template std::pair<NDArray, NDArray>
+COOGetRowDataAndIndices<kDLCPU, int32_t, int32_t>(COOMatrix, int64_t, bool);
+template std::pair<NDArray, NDArray>
+COOGetRowDataAndIndices<kDLCPU, int64_t, int64_t>(COOMatrix, int64_t, bool);
+
+///////////////////////////// CSRGetData /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType, typename DType>
+NDArray COOGetData(COOMatrix coo, int64_t row, int64_t col) {
+  CHECK(COOHasData(coo)) << "missing data array";
+  // TODO(minjie): use more efficient binary search when the column indices is sorted
+  CHECK(row >= 0 && row < coo.num_rows) << "Invalid row index: " << row;
+  CHECK(col >= 0 && col < coo.num_cols) << "Invalid col index: " << col;
+  std::vector<DType> ret_vec;
+  const IdType* coo_row_data = static_cast<IdType*>(coo.row->data);
+  const IdType* coo_col_data = static_cast<IdType*>(coo.col->data);
+  const DType* data = static_cast<DType*>(coo.data->data);
+  for (IdType i = 0; i < coo.row->shape[0]; ++i) {
+    if (coo_row_data[i] == row && coo_col_data[i] == col)
+      ret_vec.push_back(data[i]);
+  }
+  return NDArray::FromVector(ret_vec, coo.data->dtype, coo.data->ctx);
+}
+
+template NDArray COOGetData<kDLCPU, int32_t, int32_t>(COOMatrix, int64_t, int64_t);
+template NDArray COOGetData<kDLCPU, int64_t, int64_t>(COOMatrix, int64_t, int64_t);
+
+///////////////////////////// COOGetDataAndIndices /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType, typename DType>
+std::vector<NDArray> COOGetDataAndIndices(
+    COOMatrix coo, NDArray rows, NDArray cols) {
+  CHECK(COOHasData(coo)) << "missing data array";
+  // TODO(minjie): more efficient implementation for matrix without duplicate entries
+  // TODO(minjie): more efficient implementation for sorted column index
+  const int64_t rowlen = rows->shape[0];
+  const int64_t collen = cols->shape[0];
+
+  CHECK((rowlen == collen) || (rowlen == 1) || (collen == 1))
+    << "Invalid row and col id array.";
+
+  const int64_t row_stride = (rowlen == 1 && collen != 1) ? 0 : 1;
+  const int64_t col_stride = (collen == 1 && rowlen != 1) ? 0 : 1;
+  const IdType* row_data = static_cast<IdType*>(rows->data);
+  const IdType* col_data = static_cast<IdType*>(cols->data);
+
+  const IdType* coo_row_data = static_cast<IdType*>(coo.row->data);
+  const IdType* coo_col_data = static_cast<IdType*>(coo.col->data);
+  const DType* data = static_cast<DType*>(coo.data->data);
+
+  std::vector<IdType> ret_rows, ret_cols;
+  std::vector<DType> ret_data;
+
+  for (int64_t i = 0, j = 0; i < rowlen && j < collen; i += row_stride, j += col_stride) {
```

Comment:
OMP can also be used here.
It seems you support broadcast row/col cases as well as normal case here. Suggest to have some code comments here for future maintenance.

Reply:
I'm pushing elements to a vector and I don't know how many entries to push beforehand.

Reply:
i c. You have to handle broadcast here. So you cannot decide the size easily.
You can use vector<vector<data>> for omp

Reply:
And I need to concatenate them afterwards, which involves another copy from different arrays.  I wonder how much gain could I get (especially for small graphs where COO implementations mainly take place).